---
title: "AmpSeqR Introduction"
author: 
  - Jiru Han:han.ji@wehi.edu.au
  - Jacob E. Munro:munro.j@wehi.edu.au
  - Melanie Bahlo:bahlo@wehi.edu.au
date: "`r Sys.Date()`"
output: github_document
---

## `AmpSeqR`: an R package for amplicon deep sequencing studies

**AmpseqR** is an R package for analysis of amplicon deep sequencing (AmpSeq) data generated on the Illumina platform. The pipeline offers various useful functions including `Data pre-processing`, `Amplicon sequence variant (ASVs) estimation`, `Data post-processing`, and `Data visualization`. Additionally, AmpseqR includes several parameters to filter noise reads and improve the accuracy of the detected haplotype.

- **Inputs** are the standard paired-end FASTQ format provided by the common Illumina sequencing platforms (e.g., MiSeq), as well as sample barcodes and target amplicon details.

- **Outputs** are amplicon sequence variant (ASV) table, which records the number of times each exact amplicon sequence variant was observed in each sample. Additionally the pipeline will generate a summary report.

This document outlines the data format and key functions used by `AmpSeqR` for processing amplicon deep sequencing data.

### Installation

The `AmpSeqR` currently available to install from Github:

```{r}
# install using devtools packages
devtools::install_github("bahlolab/AmpSeqR")
```

### Key parameters in `AmpSeqR`

Function | Default | Description
-----|---|-------------
min_read_count | 1000 | The minimum number of reads per sample per marker
min_marker_count | 100 | The minimum number of reads for each amplicon marker after calling haplotype
n_sample | 10000 | Downsamples an exact number of reads from paired-end fastq files
min_overlap | 10 | The minimum length of the overlap required for merging the forward and reverse reads
min_asv_count | 5 | The minimum number of reads for each haplotype
min_asv_freq | 0.001 | The minimum haplotype frequency
min_ident | 0.75 | The minimum sequence similarity
min_ident_z | -3 | The minimum standardized sequence similarity
max_breakpoints | 3 | The maximum number of breakpoints chimera reads
min_parent_ratio | 1.5 | The minimum ratio that the sequences greater than this value more abundant than a sequence can be its "parents"
max_marker_miss | 0.5 | The maximum fraction of missing data for amplicon markers
max_sm_miss | 0.5 | The maximum fraction of missing data for samples
min_homo_rep | 3 | The minimum length of the homopolymer repeats
sample_med_He | 0| The maximum median expected heterozygosity for the variance of all samples in the dataset
n_alleles | 3 | The maximum number of alleles at a locus
Var_MAF | 0.001 | The minimum minor allele frequency of the variant
Var_He | 0.001 | The minimum expected heterozygosity of the variant

### Data Format

**Case study: The synthetic malaria parasite _Plasmodium falciparum_ dataset (3D7 and Dd2)** 

The dataset consists of 5 _P. falciparum_ samples with defined mixtures (1:1, 1:10, 1:100, 1:500, 1:1000) of two _P. falciparum_ reference genomes Pf3D7 and PfDd2. We selected two amplicon markers, the _P. falciparum_ surface marker genes circumsporozoite protein (*_CSP_*) and thrombospondin-related anonymous protein (_TRAP_).

```{r}
library(AmpSeqR)
example_data <- get_ampseqr_example_data()
example_data$sample_manifest
example_data$marker_info
```

- The sample barcodes file should include sample_id, barcode_fwd, barcode_rev, sample (sample name, can be the same as sample_id), info (e.g., sample type).  

- The target amplicon detail file should include marker_id, primer_fwd, primer_rev, seq (reference sequence), chrom (chromosome), start (reference sequence start position), end (reference sequence end position). 

These files should be set up in a text editor or excel and saved as text CSV format (*.csv) file.

### Run AmpSeqR

### Data pre-processing

The data pre-processing step demultiplexes the raw paired-end FASTQ reads and trims the sample barcodes and target amplicon primer sequences as well as assigning the sequence to each sample and marker (amplicon target) combination.

```{r}
# Data
reads_1 <- example_data$reads_1
reads_2 <- example_data$reads_2
sample_manifest <- example_data$sample_manifest
marker_info <- example_data$marker_info
# Set working directory
setwd("PATH_TO_WORKING_DIRECTORY")
# Create the analysis directory
dir.create("./runs")
run_dir <- "runs"
```

**Demultiplex reads**

Demultiplex the reads using the `demultiplex_reads` function as follows:

```{r}
# Demultiplex reads
demultiplexed <- demultiplex_reads(sample_manifest = sample_manifest,
                                        marker_info = marker_info,
                                        reads_1 = reads_1,
                                        reads_2 = reads_2,
                                        output_dir = run_dir,
                                        output_sub_dir = file.path(run_dir, 'demultiplex')
                                   )

demultiplexed
```

The main file outputs for the function `demultiplex_reads` are:

- demultiplex folder: demultiplexed paired-end fastq files.
-	demultiplex.rds: the demultiplexed table in RDS format which includes sample_id, marker_id, reads_1 (the forward read fastq file path), reads_2 (the reverse read fastq file path), n (number of demultiplexed reads), sample, info.

**Visualize the quality profiles of the forward and reverse reads**

Before filtering and trimming the paired-end fastq files we can visualize the quality profiles of the forward and reverse reads of different amplicon markers by using the `plot_quality` function.

```{r}
# Visualize the quality profiles of the forward and reverse reads
read_quality <- plot_quality(demultiplexed)
read_quality
```

The red line shows the mean quality score of all samples in each cycle. Color represents the proportion of reads of each quality score in each cycle.

According to the quality profiles of the forward and reverse reads, we can define the trim position for the forward and reverse reads to remove low-quality bases. The marker_trim should be a data frame with three columns marker_id (character), trim_fwd (integer), and trim_rev (integer), or can be left unset (NULL).

**Filter and trim reads**

Filter the poor-quality reads and trim low-quality bases using the `dada_filter` function as follows:

```{r}
# Filter and trim reads
flt_reads <- demultiplexed %>% 
  dada_filter(output_dir = run_dir,
              output_sub_dir = file.path(run_dir, 'filter'))
flt_reads
```

The main file outputs for the function `dada_filter` are:

- filter folder: filtered and trimmed paired-end fastq files.
-	filtered_reads.rds table: the filtered table in RDS format which includes sample_id, marker_id, n (number of demultiplexed reads), sample, info, reads_1 (the filtered and trimmed forward fastq file path), reads_2 (the filtered and trimmed reverse fastq file path), n_in (number of reads before filtering and trimming), n_out (number of reads after filtering and trimming).

**Downsampling reads**

Downsampling reads using the `downsample_reads` function as follows:

```{r}
# Downsampling reads
sub_reads <- flt_reads %>% 
  downsample_reads(output_dir = run_dir,
                   output_sub_dir = file.path(run_dir, 'downsample'),
                   min_read_count = 1000,
                   n_sample = 10000,
                   count_col = 'n_out') 
sub_reads
```

The main file outputs for the function `downsample_reads` are:

- downsample folder: downsampled paired-end fastq files with a read number greater than max_read_count.
-	subsampled_reads.rds table: the downsampled table in RDS format includes sample_id, marker_id, sample, info, reads_1 (the downsampled forward fastq file path), reads_2 (the downsampled reverse fastq file path), n_out (number of reads after filtering and trimming), n (number of reads after downsampling).

### Amplicon sequence variant (ASVs) estimation

The amplicon sequence variants (ASVs) estimation process identifies amplicon haplotype sequences. 

**Construct amplicon sequence variant table (ASV) table**

```{r}
# Construct amplicon sequence variant table (ASV) table
seq_tbl <- sub_reads %>% 
  dada_seq_tbl(output_dir = run_dir) 
seq_tbl
```

The main file outputs for the function `dada_seq_tbl` are:

-	seq_tbl.rds table: the seq_tbl table in RDS format includes sample_id, marker_id, sequence (haplotype sequence), count (haplotype read counts), status (haplotype status), sample, info.

### Data post-processing

This step primarily optimizes amplicon haplotype calling, such as indel errors, chimeric reads and variant filtering. 

Several parameters are used to filter haplotypes, such as: 

- haplotype frequency, 
- sequence similarity (haplotype sequences were mapped against reference sequences for each amplicon marker to calculate the sequence similarity), 
- variant heterozygosity, 
- minor allele frequency in a given dataset. 

**Amplicon haplotype inference**

Haplotype inference using the `annotate_seq_tbl` function as follows:

```{r}
# Haplotype inference
seq_ann_tbl <- seq_tbl %>% 
  annotate_seq_tbl(marker_info = marker_info,
                   output_dir = run_dir,
                   min_marker_count = 500,
                   min_asv_count = 5,
                   min_asv_freq = 0.001,
                   min_ident = 0.75,
                   min_ident_z = -3,
                   max_breakpoints = 3,
                   min_parent_ratio = 1.5
                   )
seq_ann_tbl
```

The main file outputs for the function `annotate_seq_tbl` are:

-	seq_ann_tbl.rds table: the annotate_seq_tbl table in RDS format includes sample_id, marker_id, sequence (haplotype sequence), count (haplotype read counts), status (haplotype status after annotation), sample, info, ident (sequence similarity), ident_z (standardized sequence similarity).

The amplicon sequence haplotype status includes various data types: 

-	**pass**, a boolean variable that describes whether the read is considered as a valid haplotype sequence, having passed all quality control checks, 
-	low_sample_count, **failed**: the sample read count lower than the minimum number of read per sample per marker, 
-	low_asv_count, **failed**: the amplicon sequence haplotype count lower than the minimum number of read for each haplotype, 
-	low_asv_freq,	**failed**: the amplicon sequence haplotype frequency lower than the minimum haplotype frequency, 
-	low_ident, **failed**: the amplicon sequence haplotype similarity (after being mapped to the reference) is lower than the minimum sequence similarity, 
-	low_ident_z, **failed**: the standardized amplicon sequence haplotype similarity (map to reference) lower than the minimum standardized sequence similarity, 
-	chimera, **failed**: the sequence is a chimera of reads, 
-	multiple, **failed**: multiple fail types.

**Haplotype filtering**

Haplotype filtering using the `sequence_filter` function as follows:

```{r}
# Haplotype filtering
seq_flt_tbl <- sequence_filter(
  seq_ann_tbl = seq_ann_tbl, 
  sample_manifest = sample_manifest,
  marker_info = marker_info, 
  output_dir = run_dir,
  vcf_output_dir = file.path(run_dir, 'vcf'), 
  min_homo_rep = 3,
  terminal_region_len = 1,
  max_sm_miss = 0.5, 
  max_marker_miss = 0.5, 
  sample_med_He = 0.1, 
  n_alleles = 3, 
  var_maf = 0.001, 
  var_he = 0.001
)
seq_flt_tbl
```

The main file outputs for the function `sequence_filter` are:

-	the seq_flt_tbl.rds table: the final amplicon haplotype sequence table in RDS format which includes sample_id, marker_id, sequence (haplotype sequence), count (read counts), masked (TRUE indicates the sequence contains variants that might be errors, and replaces the nucleotide with the reference genome nucleotide), status (haplotype status, all pass), ident (sequence similarity), haplotype (a named panel of haplotypes for each marker), frequency (within-sample haplotype frequency for each marker), sample, info.

### Data visualization

The data visualization step generates a comprehensive Rmarkdown report that contains various summary data visualizations and data summaries, such as quality checks, performed at both the individual/sample and amplicon level, amplicon haplotype sequence counts, haplotype diversity metrics of amplicon markers in the dataset, haplotype sequence visualization, haplotype tree generation, and principal component analysis. 

```{r}
# Data visualization
generate_report(sample_manifest = sample_manifest,
                marker_info = marker_info, 
                demultiplexed = demultiplexed, 
                flt_reads = flt_reads, 
                sub_reads = sub_reads,
                seq_ann_tbl = seq_ann_tbl, 
                seq_flt_tbl = seq_flt_tbl, 
                report_output_dir = file.path(run_dir, 'report'))
  
```

The main file outputs for the function `generate_report` are:

-	the Report.html report: the HTML report including various summary data visualizations.

### Fast run

We also include the `process_run` function that calls all the functions, making it easy to get all results.

```{r}
# Data
reads_1 <- example_data$reads_1
reads_2 <- example_data$reads_2
sample_manifest <- example_data$sample_manifest
marker_info <- example_data$marker_info
# Set working directory
setwd("PATH_TO_WORKING_DIRECTORY")
# Create the analysis directory
process_run_dir <- "runs_process"
dir.create("./runs_process")
```

```{r}
# Fast run
process_run(reads_1 = reads_1,
            reads_2 = reads_2,
            sample_manifest = sample_manifest,
            marker_info = marker_info,
            run_dir = process_run_dir,
            min_read_count = 1000,
            n_sample = 10000,
            min_marker_count = 500,
            min_asv_count = 5,
            min_asv_freq = 0.001,
            min_ident = 0.75,
            min_ident_z = -3,
            max_breakpoints = 3,
            min_parent_ratio = 1.5,
            min_homo_rep = 3,
            terminal_region_len = 1,
            max_sm_miss = 0.5, 
            max_marker_miss = 0.5, 
            sample_med_He = 0.1, 
            n_alleles = 3, 
            var_maf = 0.001, 
            var_he = 0.001
            )

```

### Citation

https://github.com/bahlolab/AmpSeqR

